apps/sms/views/shared/js/navigation.js

```diff
@@ -1,361 +1,921 @@
+    name: 'thread-list',
+    url: 'index.html',
+    behavior: 'InboxView',
+    alsoContains: ['thread', 'composer', 'group-view', 'report-view']
+  }),
+  Object.freeze({
+    name: 'thread',
+    behavior: 'ConversationView',
+    previous: 'thread-list',
+    partOf: 'thread-list'
+  }),
+  Object.freeze({
+    name: 'composer',
+    behavior: 'ConversationView',
+    previous: 'thread-list',
+    partOf: 'thread-list'
+  }),
+  Object.freeze({
+    name: 'group-view',
+    behavior: 'GroupView',
+    previous: 'thread',
+    partOf: 'thread-list'
+  }),
+  Object.freeze({
+    name: 'report-view',
+    behavior: 'ReportView',
+    previous: 'thread',
+    partOf: 'thread-list'
+  })
+]);
+
+const VIEWS = isUsingOldArchitecture() ? OA_VIEWS : NGA_VIEWS;
+
+var currentState;
+var navigationTransition;
+
+/**
+ * Navigate to this URL.
+ *
+ * @param {String} url URL to load.
+ */
+function setLocation(url) {
+  debug('Setting location', url);
+  window.location.assign(url);
+}
 
-// This function is used as a rejection handler, to swallow errors and log them
-function catchError(e) {
-  if (e) {
+/**
+ * We're using the old architecture if we're not in a split document.
+ *
+ * @returns {Boolean} true if we're using the old architecture.
+ */
+function isUsingOldArchitecture() {
+  debug('isUsingOldArchitecture()');
+  var pathname = window.location.pathname;
+  var result = (pathname === '/' || pathname === '/index.html');
+  debug('isUsingOldArchitecture() =>', result);
+  return result;
+}
+
+/**
+ * Parses hash to find a view name.
+ * It takes the part of the hash between '/' or the start of the string, and '?'
+ * or the end of the string.
+ *
+ * Examples:
+ *   #/conversation?id=5               => conversation
+ *   #/composer?recipient=+33615647897 => composer
+ *   #composer                          => composer
+ *
+ * @param {String} hash The string to analyze.
+ * @returns {String} The analyzed string. It is not guaranteed to be an existing
+ * view name.
+ */
+function findViewNameFromHash(hash) {
+  debug('findViewNameFromHash(%s)', hash);
+  if (!hash) {
+    return null;
+  }
+
+  if (hash[0] === '#') {
+    hash = hash.slice(1);
+  }
+
+  var index = hash.indexOf('/');
+  if (index !== -1) {
+    hash = hash.slice(index + 1);
+  }
+
+  index = hash.indexOf('?');
+  if (index !== -1) {
+    hash = hash.slice(0, index);
+  }
+
+  return hash;
+}
+
+/**
+ * Tries to find the correct view from the location.
+ *
+ * First it tries to find the view using the location's hash.
+ * If findViewNameFromHash returns something, we try to match an existing view
+ * from this name, then we check whether the found view is indeed part of this
+ * document using the full URL.
+ *
+ * If findViewNameFromHash returns nothing, we try to find a view from the full
+ * URL.
+ *
+ * @param {Location} location The content of window.location.
+ * @returns {Object} The found view, or null.
+ */
+function findViewFromLocation(location) {
```

Comment:
__note:__ in some methods like ```setLocation``` or ```isUsingOldArchitecture``` you access ```location``` from global scope, but in this method you pass it as parameter :) Is it just for consistency with ```findViewFromName``` and ```findViewFromHash```?